src/privatesend-client.cpp

```diff
@@ -1175,57 +1167,65 @@ void CPrivateSendClientManager::ProcessPendingDsaRequest(CConnman& connman)
 
 bool CPrivateSendClientSession::SubmitDenominate(CConnman& connman)
 {
+    // This is just a local helper
+    auto GetStartRound = [](bool fMixLowest, bool fScanFromTheMiddle) -> int
+    {
+        if (fScanFromTheMiddle) {
+            return privateSendClient.nPrivateSendRounds / 2;
+        } else if (!fMixLowest) {
+            return privateSendClient.nPrivateSendRounds - 1;
+        }
+        return 0;
+    };
+
     LOCK2(cs_main, pwalletMain->cs_wallet);
 
     std::string strError;
-    std::vector<CTxDSIn> vecTxDSInRet;
-    std::vector<CTxOut> vecTxOutRet;
+    std::vector< std::pair<CTxDSIn, CTxOut> > vecPSInOutPairs, vecPSInOutPairsTmp;
+
+    if (!SelectDenominate(strError, vecPSInOutPairs)) {
+        LogPrintf("CPrivateSendClientSession::SubmitDenominate -- SelectDenominate failed\n");
+        return false;
+    }
+
     // lean towards "highest" branch but still mix via "lowest" one someties
     bool fMixLowest = privateSendClient.nLiquidityProvider || (GetRandInt(4) == 0);
+    // Try to use only inputs with the same number of rounds, from low to high, or vice versa
+    int nLoopStep = fMixLowest ? 1 : -1;
     // lean towards edges but still mix starting from the middle someties
     // Note: liqudity providers always start from 0
     bool fScanFromTheMiddle = (privateSendClient.nLiquidityProvider == 0) && (GetRandInt(4) == 0);
+    // Try to mix more than one input by default. Will fallback to a single input
+    // if mixing more inputs is impossible atm.
+    // Note: liqudity providers don't care
+    bool fAvoidSingle = privateSendClient.nLiquidityProvider == 0;
 
-    int nRoundStart{0};
-    if (fScanFromTheMiddle) {
-        nRoundStart = privateSendClient.nPrivateSendRounds / 2;
-    } else if (!fMixLowest) {
-        nRoundStart = privateSendClient.nPrivateSendRounds;
-    }
+    int nRoundStart = GetStartRound(fMixLowest, fScanFromTheMiddle);
+    int nRoundEdge = GetStartRound(fMixLowest, false);
+    int nRoundStartTmp{0};
 
     // Submit transaction to the pool if we get here
-    if (fMixLowest) {
-        // Try to use only inputs with the same number of rounds, from low to high
+    while (true) {
+        nRoundStartTmp = nRoundStart;
         while (true) {
-            for(int i = nRoundStart; i < privateSendClient.nPrivateSendRounds; i++) {
-                if(PrepareDenominate(i, i + 1, strError, vecTxDSInRet, vecTxOutRet)) {
+            for (int i = nRoundStartTmp; i >= 0 && i < privateSendClient.nPrivateSendRounds; i += nLoopStep) {
+                if (PrepareDenominate(i, i + 1, strError, vecPSInOutPairs, vecPSInOutPairs, fAvoidSingle)) {
```

Comment:
...  vecPSInOutPairs, vecPSInOutPairsTmp, ...
Looks like you need to run new tests ;)

Reply:
😬 Good catch! I believe it's an issue because of solving merge conflicts due to rebase and I compiled it with correct version but worth mixing again I guess :)

Reply:
IMHO, just to remain consistent [this condition](https://github.com/dashpay/dash/blob/f17f1185e13b2ff3752e68313845a1fc00578a69/src/wallet/wallet.cpp#L3191) should be also changed to the same logic (`nRounds >= nPrivateSendRoundsMax` -> `nRounds > nPrivateSendRoundsMax`, and in call of [SelectCoinsDark](https://github.com/dashpay/dash/blob/f17f1185e13b2ff3752e68313845a1fc00578a69/src/privatesend-client.cpp#L1069) change to `privateSendClient.nPrivateSendRounds - 1`).

Reply:
Good point. And it deserves its own PR imo, pls see #2277